Plugin for SAS4A/SASSY-1

zhieejhia93 opened the pull request:
This PR adds a new plugin for SAS4A/SASSY-1 (SAS):

- The SAS plugin is similar to the MOOSE plugin. The only major difference is that SAS requires an additional step in `postrun` to convert the SAS-generated '.dat' files to 'csv' files. 
- A simple steady-state single-pin example problem is added (currently working on a more comprehensive example problem).
- A unit test is added based on the MOOSE unit test.
- The documentation is updated to include a short description for the SAS plugin.
- The SAS **INCLUDE** directive that allows a model to be made up of several different files in various subdirectories is not yet implemented.

zhieejhia93 committed:
Added SAS plugin

zhieejhia93 committed:
Minor changes to SAS example7a

zhieejhia93 committed:
Added SAS plugin to documentation.

nstauff commented on examples/example7a_SAS/example7a.py:
```diff
@@ -0,0 +1,49 @@
+sas_plugin.sas_exec = sas_dir + "/bin/sas-5.5-Darwin-x86_64.x" # SAS executable
+sas_plugin.conv_channel = sas_dir + "/Tools/macOS/CHANNELtoCSV-Darwin-x86_64.x" # SAS utility to convert "CHANNEL.dat" files to ".csv"
+sas_plugin.conv_primar4 = sas_dir + "/Tools/macOS/PRIMAR4toCSV-Darwin-x86_64.x" # SAS utility to convert "PRIMAR4.dat" files to ".csv"
+sas_result = sas_plugin.workflow(params)
```
I think we may want also to give the option for a user to provide these parameters (sam_exec, conv_channel, conv_prim4) in their bashrc to make the integration consistent with other plugins.

nstauff replied:
If we do this, we may "force" the user to provide SAS_DIR that will contain `sas.x`, `CHANNELtoCSV.x` and `PRIMAR4toCSV.x`

nstauff replied:
Thanks for the modification. I think these 3 lines can now be removed.
```
 sas_plugin.sas_exec = sas_dir + "/sas.x" # SAS executable
 sas_plugin.conv_channel = sas_dir + "/CHANNELtoCSV.x" # SAS utility to convert "CHANNEL.dat" files to ".csv"
 sas_plugin.conv_primar4 = sas_dir + "/PRIMAR4toCSV.x" # SAS utility to convert "PRIMAR4.dat" files to ".csv"
```

zhieejhia93 replied:
@nstauff I don't think I am following you here. If we remove the paths to `sas_exec` and the two utilities then the code wouldn't run, right? My understanding is that in the environment, "SAS_DIR" is only set to the directory containing these three files so the specific paths to the files still need to specified explicitly.

nstauff replied:
This is right with the current implementation. Please check the `plugin_pyarc` and `example1c` for another example (different from MOOSE). On the MOOSE plugin, we need to provide the MOOSE_DIR and the executable because we are specifying which MOOSE app needs to be run. On the PyARC, we can directly provide link to PyARC directory with PYARC_DIR, and then we don't need to specify anymore the executable (but we still have the option). 
Maybe I am overthinking it and what you have will be good. I will be interested to hear the review from @paulromano!

zhieejhia93 replied:
Thanks for the clarification. I will update the PR to follow PyARC's implementation. I think it's better this way than the current implementation.

nstauff commented on doc/source/user/usage.rst:
```diff
@@ -212,6 +212,36 @@ The path to PyARC directory must be specified explicitly with the
+    sas_plugin.conv_primar4  = "/path/to/SAS_primar4_utility"
+
+To execute PyARC, the :meth:`~watts.PluginSAS.workflow` method is called
+the same way as other Plugins.
```
To execute SAS - not PyARC

nstauff commented on doc/source/user/usage.rst:
```diff
@@ -212,6 +212,36 @@ The path to PyARC directory must be specified explicitly with the
+
+The path to the SAS4A/SASSY-1 executable must be specified explicitly with the :attr:`~watts.PluginSAS.sas_exec` attribute::
+
+    sas_plugin.sas_exec  = "/path/to/SAS"
```
I think specifying the input explicitly may not be required - we could also provide just the `SAS_DIR` in the bashrc.

nstauff commented on examples/README.md:
```diff
@@ -18,3 +18,5 @@ Training on WATTS is enabled with a suite of simple examples:
 * [__example5__](example5_multiapps/README.md) - Simple MOOSE MultiApp calculation
 ## Example 6
 * [__example6__](example6_adv_multiapps/README.md) - Advance MOOSE MultiApp calculation
+## Example 7
```
We may want to discuss at our next meeting about numbering logic for the different examples, now that we have a lot of them...

nstauff commented on src/watts/plugin_sas.py:
```diff
@@ -0,0 +1,180 @@
+
+    Parameters
+    ----------
+    template_file
```
I don't think n_cpu is needed for SAS.

nstauff commented on examples/example7a_SAS/example7a.py:
```diff
@@ -0,0 +1,49 @@
+# SPDX-FileCopyrightText: 2022 UChicago Argonne, LLC
+# SPDX-License-Identifier: MIT
+
+#########################################################
```
We may want also at the next meeting to discuss the documentation of the example. I like your approach, but it makes the README irrelevant. Let's discuss which approach implement for all cases.

paulromano replied:
If we do keep this, I would make it a proper docstring:
```Python
"""
Example problem ...
"""
```

zhieejhia93 committed:
Updated SAS plugin according to Nicolas' comments

zhieejhia93 committed:
Updated SAS plugin according to Nicolas' comments

zhieejhia93 commented:
@nstauff I updated the SAS plugin according to your comments:
- Users can now set the SAS directory as `SAS_DIR` in their environment.
- Added a second example of a simple sodium loop that uses both the core (channel) and the PRIMAR-4 modules
- I kept the header of the examples the same for now. We can talk about this in the meeting.
- Some updates on the documentation.

nstauff commented on doc/source/user/usage.rst:
```diff
@@ -212,6 +212,38 @@ The path to PyARC directory must be specified explicitly with the
+Furthermore, the paths to the SAS utilities that convert the ".dat" files to ".csv" files must be specified with the :attr:`~watts.PluginSAS.conv_channel` and :attr:`~watts.PluginSAS.conv_primar4` attributes::
+
+    sas_plugin.conv_channel  = "/path/to/SAS_channel_utility"
+    sas_plugin.conv_primar4  = "/path/to/SAS_primar4_utility"
```
we should provide here the default names of the files should be `PRIMAR4toCSV.x` and `CHANNELtoCSV.x`

zhieejhia93 committed:
Added default paths to SAS and utilities exec

zhieejhia93 commented:
@nstauff I updated the PR according to your comments:
- Updated the utilities in the documentation to their actual names.
- Added default paths to the SAS exec and utilities. Users no longer need to explicitly set these paths, as long as "SAS_DIR" is added to their environment.

paulromano commented on doc/source/user/usage.rst:
```diff
@@ -212,6 +212,38 @@ The path to PyARC directory must be specified explicitly with the
 To execute PyARC, the :meth:`~watts.PluginPyARC.workflow` method is called
 the same way as other Plugins.
 
+SAS4A/SASSY-1 Plugin
+~~~~~~~~~~~~~
```
```suggestion
SAS4A/SASSY-1 Plugin
~~~~~~~~~~~~~~~~~~~~
```

paulromano commented on examples/example7a_SAS/example7a.py:
```diff
@@ -0,0 +1,44 @@
+params['sas_version'] = 5.5
+params['tmax'] = 400.0 # maximum problem time in s
+params['flow_per_pin'] = 0.15 # kg/s
+params['total_reactor_power'] = Quantity(182, "MW") # 182000000.0 W
```
When we're specifying a quantity explicitly, I don't think we need to put a comment afterward -- it should be clear enough as is.

paulromano commented on examples/example7b_SAS/example7b.py:
```diff
@@ -0,0 +1,74 @@
+# SPDX-FileCopyrightText: 2022 UChicago Argonne, LLC
+# SPDX-License-Identifier: MIT
+
+#########################################################
```
Same here: if we keep this, make it a docstring

paulromano commented on examples/example7b_SAS/example7b.py:
```diff
@@ -0,0 +1,74 @@
+params['sas_version'] = 5.5
+params['tmax'] = 1000.0 # maximum problem time in s
+params['flow_per_pin'] = 0.15 # kg/s
+params['total_reactor_power'] = Quantity(20, "kW") # 200000.0 W
```
```suggestion
params['total_reactor_power'] = Quantity(20, "kW")
```

paulromano commented on src/watts/plugin_sas.py:
```diff
@@ -0,0 +1,203 @@
+import time
+from typing import List, Optional
+
+import os
+import glob
+import numpy as np
+import pandas as pd
+import platform
```
`os`, `glob`, and `platform` belong in the first block of imports. According to PEP8, there should be three blocks: 1) standard-library imports, 2) third-party module imports, and 3) local imports

paulromano commented on src/watts/plugin_sas.py:
```diff
@@ -0,0 +1,203 @@
+    def stdout(self) -> str:
+        return (self.base_path / "SAS_log.txt").read_text()
+
+    def _save_sas_csv(self) -> dict:
```
To me, the word "save" means we should be "saving" something (to disk). I would suggest renaming this to `_get_sas_csv_data` or something along those lines

paulromano commented on src/watts/plugin_sas.py:
```diff
@@ -0,0 +1,203 @@
+        # Check OS to make sure the extension of the executable is correct.
+        # Linux and macOS have different executables but both are ".x".
+        # The Windows executable is ".exe". 
+        if platform.system() == 'Linux' or platform.system() == 'Darwin':
+            self._sas_exec = Path(os.environ["SAS_DIR"] + "/sas.x")
+            self._conv_channel = Path(os.environ["SAS_DIR"] + "/CHANNELtoCSV.x")
+            self._conv_primar4 = Path(os.environ["SAS_DIR"] + "/PRIMAR4toCSV.x")
+        elif platform.system() == 'Windows':
+            self._sas_exec = Path(os.environ["SAS_DIR"] + "/sas.exe")
+            self._conv_channel = Path(os.environ["SAS_DIR"] + "/CHANNELtoCSV.exe")
+            self._conv_primar4 = Path(os.environ["SAS_DIR"] + "/PRIMAR4toCSV.exe")
```
```suggestion
        sas_dir = Path(os.environ.get("SAS_DIR", ""))
        ext = "exe" if platform.system() == "Windows" else "x"
        self._sas_exec = sas_dir / f"sas.{ext}"
        self._conv_channel = sas_dir / f"CHANNELtoCSV.{ext}"
        self._conv_primar4 = sas_dir / f"PRIMAR4toCSV.{ext}"
```

paulromano commented on src/watts/plugin_sas.py:
```diff
@@ -0,0 +1,203 @@
+        # they may not be outputted per user's choice.
+        # Doesn't work with run_proc(). Use os.system() for now.
+        # Needs to convert PosixPath back to string.
+        if Path("CHANNEL.dat").is_file():
+            os.system(str(self.conv_channel) + " <CHANNEL.dat> CHANNEL.csv") 
+        if Path("PRIMAR4.dat").is_file():
+            os.system(str(self.conv_primar4) + " <PRIMAR4.dat> PRIMAR4.csv") 
```
You should use the [`subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.run) function instead of `os.system`, which is not portable.

zhieejhia93 committed:
Update doc/source/user/usage.rst

Co-authored-by: Paul Romano <paul.k.romano@gmail.com>

zhieejhia93 committed:
Update src/watts/plugin_sas.py

Co-authored-by: Paul Romano <paul.k.romano@gmail.com>

zhieejhia93 committed:
Update examples/example7b_SAS/example7b.py

Co-authored-by: Paul Romano <paul.k.romano@gmail.com>

zhieejhia93 committed:
Updated according to Paul's comments

zhieejhia93 commented on src/watts/plugin_sas.py:
```diff
@@ -185,12 +186,10 @@ def postrun(self, params: Parameters) -> ResultsSAS:
-        # Needs to convert PosixPath back to string.
         if Path("CHANNEL.dat").is_file():
-            os.system(str(self.conv_channel) + " <CHANNEL.dat> CHANNEL.csv") 
+            subprocess.run(str(self.conv_channel) + " <CHANNEL.dat> CHANNEL.csv", shell=True)
```
@paulromano I changed this to `subprocess.run()` as you requested. Initially I tried passing the arguments as a list but that didn't work as intended. The executable was able to run but somehow wasn't able to locate the "CHANNEL.dat" file. I ended up having to pass a string with `shell=True`. I tested this on macOS, Linux, Windows and it seemed to work but I am not entirely sure if this would cause other problems. Please let me know if you have any suggestions for this.

paulromano replied:
The "proper" way to do this with `subprocess` is to use the `stdin` and stdout` arguments:
```Python
with open("CHANNEL.dat", "r") as file_in, open("CHANNEL.csv", "w") as file_out:
    subprocess.run(str(self.conv_channel), stdin=file_in, stdout=file_out)
```

zhieejhia93 replied:
Thanks @paulromano. I updated the the PR according to your comment.

zhieejhia93 commented:
@paulromano I updated the PR based on your comments. Thank you.

nstauff reviewed (approved):
looks good to me. Thanks @zhieejhia93 !

zhieejhia93 committed:
Updated execution of conv_channel and conv_primar4
